wordpress-mobile/WordPress-Android
# Feature/fluxc media upload service multiple uploads reattach progress

This PR makes the multiupload capabilities of the MediaUploadService reflect on the UI when going out of the Editor and back (and images in a Post are still being uploaded).

### TEST A (plain old single file upload):

0. open the editor for a new post.
1. upload one file, choosing it from the media picker (select from gallery)
2. wait till it is finished uploading
3. the post should be changed and updated in the DB, the image should unblur.
4. get out of the editor
5. get back in
6. the image should load from the web (see how it slowly forms, on Aztec)


### TEST B (background upload):

0. open the editor for a new post.
1. start uploading one file, choosing it from the media picker (select from gallery)
2. while it is uploading (you can see the progress bar), just get out of the editor
3. wait for a minute or so (whatever you think is enough for the upload to complete)
4. get back in the editor
5. the image should load from the web (see how it slowly forms, on Aztec)


### TEST C (background upload - reattach progress):

0. open the editor for a new post.
1. upload one file, choosing it from the media picker (select from gallery)
2. while it is uploading (you can see the progress bar), just get out of the editor
3. wait for a few seconds (just enough for the upload to progress a bit more, but not as much as to let it complete).
4. get back in the editor
5. the progress bar should update itself to wherever it is it’s being uploaded
6. wait till it is finished uploading
7. the post should be changed and updated in the DB, the image should unblur.
8. get out of the editor
9. get back in
10. the image should load from the web (see how it slowly forms, on Aztec)

### TEST D (multiple upload):

0. open the editor for a new post.
1. upload one file, choosing it from the media picker (select from gallery)
2. while it is uploading (you can see the progress bar), pick another file to insert
3. both files should be updating their progress, independently
4. wait till both finish and both present the same behaviour as per TEST A.


### TEST E (background upload - multiple upload):

0. open the editor for a new post.
1. upload one file, choosing it from the media picker (select from gallery)
2. while it is uploading (you can see the progress bar), pick another file to insert
3. both files should be updating their progress, independently
4. while they’re still uploading (you can see the progress bar), just get out of the editor
5. wait for a minute or so (whatever you think is enough for the upload of both files to complete)
6. get back in the editor
7. the images should load from the web (see how they slowly form, on Aztec)


### TEST F (background upload - multiple upload - reattach progress):

0. open the editor for a new post.
1. upload one file, choosing it from the media picker (select from gallery)
2. while it is uploading (you can see the progress bar), pick another file to insert
3. both files should be updating their progress, independently
4. while they’re still uploading (you can see the progress bar), just get out of the editor
5. wait for a few seconds (just enough for the uploads to progress a bit more, but not as much as to let both complete).
6. get back in the editor
7. the progress bar should update itself to wherever it is it’s being uploaded
8. wait till both are finished uploading (both images unbar)
9. get out of the editor, then get back in
10. the images should load from the web (see how they slowly form, on Aztec)

### TEST G (multicancel):

0. open the editor for a new post.
1. upload one file, choosing it from the media picker (select from gallery)
2. while it is uploading (you can see the progress bar), pick another file to insert
3. both files should be updating their progress, independently
4. while they’re still uploading (you can see the progress bar), pick a third one
5. cancel the first one
6. now cancel the second one
7. the third one should still be uploading
8. wait till it finishes, and make sure it’s the only one finishing (wait some minutes and check the state is as expected: 2 marked for retry, third one completed).


## Known Issues

(Planning to work on these on separate PRs)

### Different working copies of PostModel
- It could happen that 2 images are being uploaded, then both finish practically at the same time, then when we try to modify the post’s content (to replace the local image with the remote one) we:
1. take the post from the db, where 2 images are local and “uploading”
2.a) image A gets replaced with the remote one, and this gets saved into the DB
2.b) image B gets replaced with the remote one, and this gets saved into the DB
We end up having a post with image A local, and image B remote, because what was saved last is 2.b) with a copy of the post that had both local images only.

How to solve this?
We need a way to ensure that all images get saved / replacing the post content.
Fortunately, the service runs as long as there’s stuff enqueued for upload.
- In the MediaUploadService, what I’m going to do now is wait for all images in the queue (for a given post) to complete, and only then a) get a copy of the post’s content and b) make all replacements at once, _then_ save.

See #5858 

### Random issues / enhancements
- would be great to get the keyboard already ready and up when creating a new post (see #5859)
- eliminate the mAztecReady flag, might not be necessary
- ~The profiling is missing the stop calls~
```
                ProfilingUtils.split("EditorFragment.onDomLoaded completed");
                ProfilingUtils.dump();
                ProfilingUtils.stop();
```
~(see #5860)~
- ~if I start a draft and do nothing, get out of the editor, it gets saved (it probably shouldn’t, should delete the post entirely). (see #5861)~
- saveContentFromSource() is not used, can be removed. (see #5864)
- when uploading several images, the progress upload state refreshes make it nearly impossible to navigate the content (scrolling up/down jumps a lot) (already tracked in https://github.com/wordpress-mobile/AztecEditor-Android/issues/313) 
- if you trash a post the snackbar is shown (and the item in the list disappears), then if you refresh the list before the snackbar is hidden, the post re-appears (because it hasn’t been deleted on the backend yet).
https://github.com/wordpress-mobile/WordPress-Android/issues/5862



cc @aforcier
